### PR TITLE
statesync: shut down node when statesync fails

### PR DESCRIPTION
**Pull request by cmwaters**
The node currently runs statesync in a separate go routine when running `OnStart()`. If statesync fails it errors silently and the node continues to run  however neither blocksync or consensus is running so the node just chills there doing nothing. This isn't desirable. If state sync fails the node should log the error and gracefully shutdown

### CODE REVIEW

**Review by tychoish — APPROVED** [review]
This seems great. 

Perhaps it'd also make sense to have a mode where nodes that couldn't start state syncing would fall back to blocksync?

**Comment by cmwaters** [comment]
> Perhaps it'd also make sense to have a mode where nodes that couldn't start state syncing would fall back to blocksync?

I thought about it but this comment here (https://github.com/tendermint/tendermint/issues/5742#issuecomment-738117783) swayed my thinking that it's better to just stop

**Comment by tychoish** [comment]
yeah, I think this is right for now, and we'd want to maybe think about the UX of config options in general before changing statesync-fallback.

File: node/node.go [hunk 1]
```diff
@@ -704,7 +704,10 @@ func (n *nodeImpl) OnStart() error {
 			if err != nil {
 				n.Logger.Error("state sync failed", "err", err)
 				// stop the node
-				n.Stop()
+				err := n.Stop()
+				if err != nil {
```
Comment (tychoish) [inline node/node.go:708]: I tend to combine these lines (also the err is already declared)

File: node/node.go [hunk 2]
```diff
@@ -703,6 +703,11 @@ func (n *nodeImpl) OnStart() error {
 			state, err := n.stateSyncReactor.Sync(context.TODO())
 			if err != nil {
 				n.Logger.Error("state sync failed", "err", err)
+				// stop the node
+				err := n.Stop()
+				if err != nil {
```
Comment (cmwaters) [inline node/node.go:708]:
```suggestion
				if err := n.Stop(); err != nil {
```

File: node/node.go [hunk 3]
```diff
@@ -703,6 +703,11 @@ func (n *nodeImpl) OnStart() error {
...
+					n.Logger.Error("error halting node", "err", err)
```
Comment (cmwaters) [inline node/node.go:709]:
```suggestion
					n.Logger.Error("failed to shut down node", "err", err)
```

File: node/node.go (hunk 1 above)
Comment (creachadair) [reply node/node.go]: Agreed, though this is the first declaration in this scope so it actually works (we shouldn't lean on that).

File: node/node.go [hunk 4]
```diff
@@ -703,6 +703,10 @@ func (n *nodeImpl) OnStart() error {
 			state, err := n.stateSyncReactor.Sync(context.TODO())
 			if err != nil {
 				n.Logger.Error("state sync failed", "err", err)
```
Comment (creachadair) [inline node/node.go:705]: Consider maybe also logging that we are shutting down the node?

```suggestion
				n.Logger.Error("state sync failed; shutting down this node", "err", err)
```

**Review by creachadair — APPROVED** [review]

**Review by williambanfield — APPROVED** [review]

**Pull request merged**